### DISCUSSION TITLE
Possibility of a pdfium processor for PDFs?

### DISCUSSION THREAD

**Discussion by cmhdave** [cantaloupe-project/cantaloupe#513] [Ideas]
We are currently using the PDFBox implementation quite extensively in Cantaloupe but memory and CPU overhead is quickly becoming an issue.  While searching around for possible solutions I came across mention of [libvips](https://libvips.github.io/libvips/) which looks promising.  Of course the immediate question for me is whether anyone had created a JNI binding for this... [they have](https://github.com/kdekooter/jlibvips).  It seems a small leap to add this to Cantaloupe.  Thoughts?

**Comment by adolski** [comment]
VIPS is a whole big image processing library. To read PDFs it uses either the PDFium, Poppler, or ImageMagick libraries. I would want to test these to see how they perform compared to PDFBox, to help me decide which one to use (if any), and then develop an Image I/O plugin that calls into it.

As far as that goes, there is a new [Foreign Linker API](https://openjdk.java.net/jeps/389) (incubating in JDK 16, targeted for 17) that should make calling into native code a lot easier compared to JNI. I expect that it will become the de facto Java standard for native interop. (There is also JNA which is slow, and JNR-FFI which is nice but another dependency.)

**Reply by cmhdave** [reply]
Funny you should mention it, I started playing around with trying to create an implementation with PDFium.  However not knowing much about it I did it with JNA (didn't know it was slower than JNI).

**Reply by cmhdave** [reply]
I finally got around to testing out my JNA wrapper around pdfium and it is much faster than PDFBox (anecdotally);  on the order of twice as fast.  I have tested some of our really big PDF examples to test with but they are images embedded in a PDF (one per page) and not born digital PDFs.  I do plan to test born digital PDFs as well.  I also want to do a more "scientific" test.

Overall it looks extremely promising if a load test doesn't turn up any memory leaks or excessive CPU usage.

**Reply by cmhdave** [reply]
One issue I'm running into with writing this as an Image I/O plugin is that the `input` for the reader comes from an `ImageInputStream`.  PDFium works with files or a document that is completely loaded into memory.  This would be fine for smaller PDFs but for the really large ones (hundreds of MB to GB) this is not an ideal scenario.  

One option I started tinkering with was saving the PDF to a scratch directory and then point PDFium at that but that'll kill performance for large PDFs again because we're copying the whole thing to file (let's say 2GB which is a filesize we've actually run into) and THEN finally processing the page we're interested in once it's stored as a file.

The beauty of your Processor model is that I can work directly with files.  I already concluded that I could only use PDFium with the `FileProcessor` rather than the `StreamProcessor`.

**Comment by adolski** [comment]
If you are intending to make this plugin publicly available, I would try to allow reading from whatever type is provided to your `ImageReaderSpi.canDecodeInput(Object)` implementation--if `Object` is a `File` or `Path`, set up PDFium for that, otherwise buffer the data (from what is probably an `ImageInputStream`) and set it up for reading from memory. The problem with `FileProcessor` is that it only works efficiently with `FilesystemSource`.

I agree that a 2 GB PDF seems large, but it may be OK for a user with enough memory. And another user may only need to serve smaller PDFs.

**Reply by cmhdave** [reply]
A lot has happened since I posted this topic.  I tried creating pdfium, JPedal, Ghostscript, iText and pdfimages processors to see if I could get better performance.  Almost all of them had similar performance to PDFBox (a couple much worse).  In the end we found that a commercial wrapper around Adobe by Datalogics was far and away the best performance (on the order of 10x).  It's not cheap so not very practical for all applications but exactly what we need.  If anyone is interested the details on how to implement is very straightforward and lifted almost directly from the Datalogics example Java code.
